x/transferinject/ibc_middleware.go

```diff
@@ -0,0 +1,193 @@
+package transferinject
+
+import (
+	"errors"
+	. "slices"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	errortypes "github.com/cosmos/cosmos-sdk/types/errors"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+
+	rtypes "github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+	"github.com/dymensionxyz/dymension/v3/x/transferinject/types"
+)
+
+type IBCMiddleware struct {
+	porttypes.IBCModule
+	porttypes.ICS4Wrapper
+
+	rollappKeeper types.RollappKeeper
+	bankKeeper    types.BankKeeper
+}
+
+// NewIBCSendMiddleware creates a new ICS4Wrapper.
+// It intercepts outgoing IBC packets and adds token metadata to the memo if the rollapp doesn't have it.
+// This is a solution for adding token metadata to fungible tokens transferred over IBC,
+// targeted at rollapps that don't have the token metadata for the token being transferred.
+// More info here: https://www.notion.so/dymension/ADR-x-IBC-Denom-Metadata-Transfer-From-Hub-to-Rollapp-d3791f524ac849a9a3eb44d17968a30b
+func NewIBCSendMiddleware(
+	ics porttypes.ICS4Wrapper,
+	rollappKeeper types.RollappKeeper,
+	bankKeeper types.BankKeeper,
+) porttypes.ICS4Wrapper {
+	return &IBCMiddleware{
+		ICS4Wrapper:   ics,
+		rollappKeeper: rollappKeeper,
+		bankKeeper:    bankKeeper,
+	}
+}
+
+// NewIBCAckMiddleware creates a new IBCModule.
+// It intercepts acknowledged incoming IBC packets and adds token metadata that had just been registered on the rollapp itself,
+// to the local rollapp record.
+func NewIBCAckMiddleware(
+	ibc porttypes.IBCModule,
+	rollappKeeper types.RollappKeeper,
+) porttypes.IBCModule {
+	return &IBCMiddleware{
+		IBCModule:     ibc,
+		rollappKeeper: rollappKeeper,
+	}
+}
+
+// SendPacket wraps IBC ChannelKeeper's SendPacket function
+func (m *IBCMiddleware) SendPacket(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	destinationPort string, destinationChannel string,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	data []byte,
+) (sequence uint64, err error) {
+	packet := new(transfertypes.FungibleTokenPacketData)
+	if err = types.ModuleCdc.UnmarshalJSON(data, packet); err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnknownRequest, "unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	if types.MemoAlreadyHasPacketMetadata(packet.Memo) {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnauthorized, "denom metadata already exists in memo")
+	}
```

Comment:
just curious why this is above the 
if rollapp == nil {

?

Reply:
well if we don't find the rollapp, it means we're sending to a non-rollapp chain, so we don't include the metadata, but that's a simplistic way to check if the other chain supports this middleware. In the future it would be nice to consider finding a nicer way to discriminate the receiving end's support for this middleware

Reply:
ok thanks